### ISSUE TITLE
Merging 2 countries fails for unknown reason

### ISSUE THREAD

**Issue by labo-flg** [LuatixHQ/opencti#6632]
## Description

This issue was reported by a community member on slack. It happens that when merging 2 countries on their platform, it fails with `Cannot read properties of undefined (reading '_index')`.

I could not reproduce this locally or on our testing environment.

## Environment

Reported initially on OCTI 5.12.33, user updated to 6.0.9 and bug still here.

## Reproducible Steps

![merge1](https://github.com/OpenCTI-Platform/opencti/assets/146674147/278679b0-c166-42b3-a10f-3a94fab15556)
![merge2](https://github.com/OpenCTI-Platform/opencti/assets/146674147/9d465729-436d-41da-a827-37941c8e7933)
![merge3](https://github.com/OpenCTI-Platform/opencti/assets/146674147/f0559948-3b28-40bd-8403-ca1b383e1c04)

## Additional information

One simple reason would have been that a task is running prior to the merge and is deleting one country ; when executing the merge it fails.
This is apparently not the case.
User has no issues accessing the countries individually.
"SA" country entity only has 3 relationships, but the full "Saudi Arabia" has 423 relationships.

Complete log on 6.0.9:

```
2024-04-05T15:56:40.307084013Z {"category":"APP","errors":[{"attributes":{"executions":[{"id":"83e6d3ca-bdc0-4bd4-a51f-4afe11f6dd3f","message":"Cannot read properties of undefined (reading '_index') - no reason provided"}],"genre":"TECHNICAL","http_status":500},"message":"Tasks execution failure","name":"UNKNOWN_ERROR","stack":"UNKNOWN_ERROR: Tasks execution failure\n    at error (/opt/opencti/build/src/config/errors.js:8:10)\n    at UnknownError (/opt/opencti/build/src/config/errors.js:76:47)\n    at executeProcessing (/opt/opencti/build/src/manager/taskManager.js:439:18)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at taskHandler (/opt/opencti/build/src/manager/taskManager.js:484:22)\n    at /opt/opencti/build/src/manager/taskManager.js:514:9\n    at kTt.#runHandlerAndScheduleTimeout (/opt/opencti/build/node_modules/set-interval-async/dist/set-interval-async-timer.cjs:36:13)\n    at Timeout._onTimeout (/opt/opencti/build/node_modules/set-interval-async/dist/set-interval-async-timer.cjs:29:13)"}],"level":"error","message":"Tasks execution failure","timestamp":"2024-04-05T15:56:40.305Z","version":"6.0.9"}
```

**Comment by nino-filigran** [comment]
Also tried on demo and was not successful at reproducing. @Kedae do you have an idea?

**Comment by UNSUB7** [comment]
If it helps, the shortened country aliases i.e. FR, SA, US, etc. were populated from VirusTotal. We turned off the VT connector a little while ago and don't expect the shorted country entities to repopulate if we're able to successfully merge them.

**Comment by SamuelHassine** [comment]
Generally, this means one of the 2 entities has been deleted before the merge background task executes.

@labo-flg please talk with @richard-julien as we already had this subject in mind, to raise a more meaningful error.

**Comment by SamuelHassine** [comment]
Not really sure but it is the first lead....

**Comment by labo-flg** [comment]
> Generally, this means one of the 2 entities has been deleted before the merge background task executes.

Yes that was the first idea, however I'm pretty confident it's not the case judging from our first discussions on slack.
If I understand correctly, we would end up with a missing entity (if one is deleted prior to the merge attempt). 
It's not the case here ; the merge fails and the 2 entities are still present.

**Comment by labo-flg** [comment]
> @labo-flg please talk with @richard-julien as we already had this subject in mind, to raise a more meaningful error.

**Comment by labo-flg** [comment]
Sorry for closing the issue, it's a missclick 😇

> @labo-flg please talk with @richard-julien as we already had this subject in mind, to raise a more meaningful error.

we can definitely improve error handling during the merge execution, in the case of one missing entity raise a descriptive error.
This would rule out or not this hypothesis for @UNSUB7's issue.

EDIT: this case is theoretically covered under the error "Cannot access all entities for merging" ; so definitely something else.

**Comment by SouadHadjiat** [comment]
Does the user asking for the merge has access to all the relationships linked to the two countries ?

**Comment by UNSUB7** [comment]
Looks like there are a number of relationships I can't view even though I'm logged in as an admin. I'm guessing the **Restricted** relationships are causing the merge error?

<img width="1360" alt="Screenshot 2024-04-25 at 11 06 18 AM" src="https://github.com/OpenCTI-Platform/opencti/assets/132004414/d7ddb9e6-772f-4c70-93b0-b69d1581acb2">

**Comment by labo-flg** [comment]
This is a good lead!
If one side appears as "restricted" it means that you do not have access to this side of the relationship for some reasons.
The main reason would be that you do not have sufficient access rights of course, but this is also the case if the object has been removed from the platform, leaving an "orphan" relationship.

It can happen if someone deletes an object in the platform. 
Doing so opencti will compute the full cluster of elements to delete = the main element + all relationships linked to it.
But if the user has not full access to all these relationships, they are not deleted, leaving orphan relationships.
We've noticed this bug lately, see #6421

@UNSUB7 I have reproduced your issue and can confirm that once you're in this state, the merge indeed throws error `Cannot read properties of undefined (reading '_index')`.

To solve it, please delete all relationships that appear as restricted on the elements you are trying to merge. 
If indeed you are admin with Bypass capability, we can assume these are orphans relationships, safe to delete.

Then retry to merge, and tell me if it's ok please.

**Comment by UNSUB7** [comment]
Deleted the restricted relationships and attempted the merge, but it still threw the same error
`Cannot read properties of undefined (reading '_index') - no reason provided`

The country entity that I'm attempting to merge was created by the same source as the "Restricted" relationships, so maybe that has something to do with the error too? The author for both entities is VirusTotal. We're tempted to just delete the entities altogether since the data is old and irrelevant at this point (data is from late 2022). If we go that route, is there a way to prevent the creation of these shortened country entities?

**Comment by labo-flg** [comment]
> The country entity that I'm attempting to merge was created by the same source as the "Restricted" relationships, so maybe that has something to do with the error too? The author for both entities is VirusTotal. We're tempted to just delete the entities altogether since the data is old and irrelevant at this point (data is from late 2022). If we go that route, is there a way to prevent the creation of these shortened country entities?

The fix merged in #6821 should prevent `Cannot read properties of undefined (reading '_index')` during merge.

When you merge 2 entities, names of the secondary entities are added to the main one as aliases.
If anyone or anything recreates an entity, the deduplication process will make sure they are not created once again as it searches in aliases too.

However, note that depending of the confidence levels of the entity and the source, it might replace the name of the entity.
Say you have Saudi Arabia with alias SA and confidence 100, if the user associated with VT connector has confidence 100 and send once again a country SA, the final name will be SA with alias Saudi Arabia (100 >= 100).
